### PR TITLE
Upgrade to support hapi v17

### PR DESCRIPTION
**Pull request by p16**
This PR updates the library, its tests and the example to run on hapi v17.

The code in this PR uses `aync`/`await` so it introduces a breaking change (nodejs >= 8).

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -1,9 +1,6 @@
 language: node_js
 sudo: false
 node_js:
-  - "4"
-  - "6"
-  - "7"
   - "8"
```
Comment (mcollina) [inline .travis.yml:4]: Can you add Node 10 here?

File: multines.js [hunk 2]
```diff
@@ -49,57 +50,60 @@ function register (server, options, next) {
...
-          return next(err)
-        }
-
-        mq.on(topic, deliverMap[path], next)
+      await wrapSubscribe(socket, path, params)
+      await new Promise((resolve, reject) => {
+        mq.on(topic, deliverMap[path], (err) => {
+          if (err) return reject(err)
+          resolve()
+        })
```
Comment (mcollina) [inline multines.js]: can you use util.promisify instead?

File: multines.js [hunk 3]
```diff
@@ -49,57 +50,60 @@ function register (server, options, next) {
...
+        return
+      }
 
-        mq.removeListener(path.replace(/^\//, ''), deliverMap[path], function () {
-          setImmediate(next)
+      await new Promise((resolve, reject) => {
+        mq.removeListener(path.replace(/^\//, ''), deliverMap[path], (err) => {
+          if (err) return reject(err)
+          resolve()
         })
```
Comment (mcollina) [inline multines.js]: Can you use util.promisify?

File: multines.js [hunk 4]
```diff
@@ -49,57 +50,60 @@ function register (server, options, next) {
...
+    await new Promise((resolve, reject) => {
+      mq.emit({
+        topic: path.replace(/^\//, ''), // the first is always a '/'
+        body
+      }, (err) => {
+        if (err) return reject(err)
+      }, (err) => {
+        if (err) return reject(err)
+        resolve()
+      })
```
Comment (mcollina) [inline multines.js]: this only have 1 callback. Can you add a `util.promisify`?

File: multines.js [hunk 5]
```diff
@@ -49,57 +50,60 @@ function register (server, options, next) {
...
-      mq.close(function () {
-        setImmediate(done)
+  server.ext('onPostStop', async () => {
+    await new Promise((resolve, reject) => {
+      mq.close((err) => {
+        if (err) return reject(err)
+        resolve()
       })
-    }
+    })
```
Comment (mcollina) [inline multines.js]: can you use util.promisify?

**Review by mcollina — COMMENTED** [review]
I haven't looked at the tests yet.

File: multines.js (hunk 2 above)
Comment (p16) [reply multines.js]: @mcollina I've tried that but it complains about `mqemitter` accessing a variable on `this` (ie: `Cannot read property '_matcher' of undefined`)

Comment (p16) [reply multines.js]: https://github.com/mcollina/mqemitter/blob/master/mqemitter.js#L76

Comment (p16) [reply multines.js]: @mcollina sorry, found what you meant. it's ugly but it does the trick

```util.promisify(mq.on.bind(mq))(topic, deliverMap[path])```

File: multines.js [hunk 6]
```diff
@@ -49,57 +51,38 @@ function register (server, options, next) {
...
 
-      wrapSubscribe(socket, path, params, (err) => {
-        if (err) {
-          return next(err)
-        }
-
-        mq.on(topic, deliverMap[path], next)
-      })
+      await wrapSubscribe(socket, path, params)
+      await util.promisify(mq.on.bind(mq))(topic, deliverMap[path])
```
Comment (mcollina) [inline multines.js]: can you just store `util.promisify(mq.on.bind(mq))` at the top?

File: multines.js [hunk 7]
```diff
@@ -49,57 +51,38 @@ function register (server, options, next) {
...
-        }
+      if (!deliverMap[path]) {
+        return
+      }
 
-        mq.removeListener(path.replace(/^\//, ''), deliverMap[path], function () {
-          setImmediate(next)
-        })
-      })
+      await util.promisify(mq.removeListener.bind(mq))(path.replace(/^\//, ''), deliverMap[path])
```
Comment (mcollina) [inline multines.js]: can we just store `util.promisify(mq.removeListener.bind(mq))` in a variable at the top?

File: test.js [hunk 8]
```diff
@@ -16,212 +16,150 @@ const expect = Code.expect
...
 
-        client.request({
-          path: '/echo',
-          method: 'POST',
-          payload: { hello: 'world' }
-        }, (err) => {
-          if (err) {
-            return done(err)
-          }
+        client.disconnect().then(resolve)
```
Comment (mcollina) [inline test.js]: Can you add a ` catch(reject)` here?

File: test.js [hunk 9]
```diff
@@ -16,212 +16,150 @@ const expect = Code.expect
...
+      }
+
+      return client.subscribe('/echo', handler)
+        .then(() => {
+          return client.request({
+            path: '/echo',
+            method: 'POST',
+            payload: { hello: 'world' }
+          })
```
Comment (mcollina) [inline test.js]: can you use await instead of then?

Comment (p16) [reply test.js]: @mcollina you mean something like the following?

```
await new Promise( async (resolve, reject) => {
  function handler (message, flags) {
    expect(message).to.equal({ hello: 'world' })

    client.disconnect().then(resolve).catch(reject)
  }
      
  try {
    await client.subscribe('/echo', handler)
    await client.request({ path: '/echo', ... })
  } catch (e) {
    reject(e)
  }
})
```

Comment (mcollina) [reply test.js]: I'm not happy with that either, forget about it. Lab@15 and async await are not great of testing this pattern.

Comment (mcollina) [reply test.js]: Side note, this does not seem correct either. It's *very hard* to test this pattern with async/await, and I would just want another testing library at this point.

Do:

```js
let resolve
let reject

  function handler (message, flags) {
    expect(message).to.equal({ hello: 'world' })

    client.disconnect().then(resolve).catch(reject)
  }

const p = new Promise((r1, r2) => {
  resolve = r1
  reject = r2
})

await client.subscribe('/echo', handler)
await Promise.all([client.request({ path: '/echo', ... }), p])
```

**Comment by p16** [comment]
@mcollina all changes done (not sure why but I still see the old code here on github, but the new commit is in with its changes 🤔 https://github.com/mcollina/multines/pull/6/commits/d2865b076eef49982d907f514a0f23bd16973b9c )

File: test.js (hunk 9 above)
Comment (mcollina) [reply test.js]: @p16 can you do the above?

File: test.js [hunk 10]
```diff
@@ -16,212 +16,156 @@ const expect = Code.expect
...
-          }, (err) => {
-            if (err) {
-              return done(err)
-            }
-
-            client2.disconnect()
+          })
+          .then(() => {
+            return client2.disconnect()
           })
```
Comment (mcollina) [inline test.js]: can you use the same pattern here?

File: test.js [hunk 11]
```diff
@@ -233,71 +177,53 @@ experiment('with shared mqemitter', () => {
...
           method: 'POST',
           payload: { hello: 'world' }
-        }, (err) => {
-          if (err) {
-            return done(err)
-          }
-
-          client.disconnect(() => setImmediate(done))
-        })
+        }).then(() => client.disconnect())
```
Comment (mcollina) [inline test.js]: can you use the same pattern here too?

**Review by mcollina — COMMENTED** [review]
Can you ignore package-lock.json and remove it from this PR?

**Comment by p16** [comment]
@mcollina removed `package-lock.json` and updated test pattern

**Review by mcollina — APPROVED** [review]
LGTM

**Pull request merged**